src/dune_pkg/rev_store.ml

```diff
@@ -2,12 +2,76 @@ open Stdune
 open Dune_vcs
 module Process = Dune_engine.Process
 module Display = Dune_engine.Display
+module Scheduler = Dune_engine.Scheduler
 module Re = Dune_re
+module Flock = Dune_util.Flock
 open Fiber.O
 
 type t = { dir : Path.t }
+
+let lock_path { dir } =
+  let parent = dir |> Path.parent_exn in
+  Path.relative parent "rev-store.lock"
+;;
+
 type rev = Rev of string
 
+(* Async-inspired variation of [Fiber.repeat_until] *)
+let rec repeat_until_finished state f =
+  let* computation = f state in
+  match computation with
+  | `Repeat state -> repeat_until_finished state f
+  | `Finished result -> Fiber.return result
+;;
+
+let attempt_to_lock flock lock ~max_tries =
+  let sleep_duration = 0.1 in
+  repeat_until_finished max_tries (function
+    | 0 -> Fiber.return @@ `Finished (Ok `Failure)
+    | retry ->
+      (match Flock.lock_non_block flock lock with
+       | Ok `Success as ok -> Fiber.return @@ `Finished ok
+       | Ok `Failure ->
+         let+ () = Scheduler.sleep sleep_duration in
+         `Repeat (retry - 1)
+       | err -> Fiber.return @@ `Finished err))
+;;
+
+let with_flock lock_path ~f =
+  let open Fiber.O in
+  let parent = Path.parent_exn lock_path in
+  Path.mkdir_p parent;
+  let fd = Unix.openfile (Path.to_string lock_path) [ Unix.O_CREAT; O_RDONLY ] 0o644 in
+  let flock = Flock.create fd in
+  let max_tries = 50 in
+  Fiber.finalize
+    ~finally:(fun () ->
+      (* closing the fd releases the flock automatically *)
+      match Unix_error.Detailed.catch Unix.close fd with
+      | Ok () ->
+        (* delete the lock to signal to the user we don't hold a lock *)
+        Fiber.return @@ Path.unlink_no_err lock_path
+      | Error detailed ->
```

Comment:
Why is this a `Code_error` and not a `User_error`?

Reply:
If there is some error that prevents the lock being released, it is probably a bug in our codebase. Like the FD is wrong or something, in any case not something that the user can do much about except for reporting it as a bug. Maybe if the lockfile was on NFS and the server went away or something.

Reply:
You can also go with `Unix_error.Detailed.raise detailed`